diff --git a/vllm_mlu/vllm_mlu/__init__.py b/vllm_mlu/vllm_mlu/__init__.py
new file mode 100644
index 000000000..d296626f3
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/__init__.py
@@ -0,0 +1,25 @@
+#
+# Copyright (c) 2025 Cambricon Technologies Co., Ltd. All Rights Reserved.
+# This file is a part of the vllm-mlu project.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+def register_mlu_platform():
+    """Register the MLU platform."""
+    return "vllm_mlu.platforms.mlu.MLUPlatform"
+
+
+def register_mlu_hijack():
+    from vllm_mlu import mlu_hijack
+    return
\ No newline at end of file

