diff --git a/flagscale/runner/runner_inference.py b/flagscale/runner/runner_inference.py
index 17f122bf..3c67459d 100644
--- a/flagscale/runner/runner_inference.py
+++ b/flagscale/runner/runner_inference.py
@@ -86,15 +86,6 @@ def _generate_run_script_inference(config, host, node_rank, cmd, background=True
 
     root_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
 
-    fs_path = sys.path.pop(0)
-    try:
-        package_path = importlib.util.find_spec("vllm").origin
-    except:
-        raise ValueError("vllm package not found.")
-    sys.path.insert(0, fs_path)
-    path = package_path.split("/")[:-2]
-    vllm_dir = "/".join(path)
-
     cmds_config = config.experiment.get("cmds", None)
     if cmds_config:
         before_start = cmds_config.get("before_start", "")
@@ -108,7 +99,7 @@ def _generate_run_script_inference(config, host, node_rank, cmd, background=True
         f.write(f"\n")
         f.write(f"cd {root_dir}\n")
         f.write(f"\n")
-        f.write(f"export PYTHONPATH={vllm_dir}:{root_dir}\n")
+        f.write(f"export PYTHONPATH={root_dir}:${{PYTHONPATH}}\n")
         f.write(f"\n")
         f.write(f'cmd="{cmd}"\n')
         f.write(f"\n")

