diff --git a/tools/checkpoint/loader_mcore.py b/tools/checkpoint/loader_mcore.py
index f885da7e..f308d1e6 100644
--- a/tools/checkpoint/loader_mcore.py
+++ b/tools/checkpoint/loader_mcore.py
@@ -203,7 +203,7 @@ def _load_checkpoint(queue, args):
     mpu.set_tensor_model_parallel_world_size(tp_size)
     mpu.set_pipeline_model_parallel_world_size(pp_size)
     mpu.set_expert_model_parallel_world_size(ep_size)
-    mpu.set_virtual_pipeline_model_parallel_world_size(vp_size)
+    mpu.set_virtual_pipeline_model_parallel_world_size(margs.virtual_pipeline_model_parallel_size)
     mpu.set_tensor_model_parallel_rank(0)
     mpu.set_pipeline_model_parallel_rank(0)
     mpu.set_expert_model_parallel_rank(0)
@@ -256,6 +256,7 @@ def _load_checkpoint(queue, args):
     md.add_bias_linear = margs.add_bias_linear
     md.add_qkv_bias = margs.add_qkv_bias
     md.norm_has_bias = margs.norm_has_bias
+    md.qk_layernorm = margs.qk_layernorm
     md.swiglu = margs.swiglu
     md.previous_num_experts = margs.num_experts
     md.previous_tensor_parallel_size = margs.tensor_model_parallel_size

