diff --git a/vllm/sampling_params.py b/vllm/sampling_params.py
index 322e53b75..f2a05db8e 100644
--- a/vllm/sampling_params.py
+++ b/vllm/sampling_params.py
@@ -210,6 +210,7 @@ class SamplingParams(
     include_stop_str_in_output: bool = False
     truncate_prompt_tokens: Optional[Annotated[int, msgspec.Meta(ge=1)]] = None
     output_kind: RequestOutputKind = RequestOutputKind.CUMULATIVE
+    guidance_scale: Optional[float] = None # --- FLAGSCALE MODIFICATION ---
 
     # The below fields are not supposed to be used as an input.
     # They are set in post_init.
@@ -563,7 +564,8 @@ class SamplingParams(
             f"{self.spaces_between_special_tokens}, "
             f"truncate_prompt_tokens={self.truncate_prompt_tokens}, "
             f"guided_decoding={self.guided_decoding}, "
-            f"extra_args={self.extra_args})")
+            f"extra_args={self.extra_args}, "
+            f"guidance_scale={self.guidance_scale})") # --- FLAGSCALE MODIFICATION ---
 
 
 class BeamSearchParams(
