diff --git a/verl/models/mcore/loader.py b/verl/models/mcore/loader.py
index 577ffc5e..6ca0bf7b 100644
--- a/verl/models/mcore/loader.py
+++ b/verl/models/mcore/loader.py
@@ -66,8 +66,8 @@ def load_state_dict_to_megatron_gptmodel(state_dict, wrapped_models, config, par
     start_time = time.time()
 
     def _get_gpt_model(model):
-        return model
-
+            return model
+    
     def broadcast_params(module):
         for param in module.parameters():
             torch.distributed.broadcast(
