diff --git a/verl/models/mcore/model_forward.py b/verl/models/mcore/model_forward.py
index b6192386..4f5fc5c3 100644
--- a/verl/models/mcore/model_forward.py
+++ b/verl/models/mcore/model_forward.py
@@ -101,6 +101,8 @@ def gptmodel_forward_no_padding(
     batch_size = input_ids.shape[0]
     input_ids_rmpad, packed_seq_params = preprocess_packed_seqs_no_padding(input_ids, pre_process=pre_process)
     input_ids_rmpad = input_ids_rmpad.contiguous()
+    
+    # if use_sequence_packing, the shape of model output is [1, total_tokens, vocab_size]
     output_orig = model(
         input_ids=input_ids_rmpad,
         attention_mask=None,
