diff --git a/vllm/distributed/device_communicators/shm_broadcast.py b/vllm/distributed/device_communicators/shm_broadcast.py
index fa944407a..2746ea0e9 100644
--- a/vllm/distributed/device_communicators/shm_broadcast.py
+++ b/vllm/distributed/device_communicators/shm_broadcast.py
@@ -422,7 +422,8 @@ class MessageQueue:
                     # we need to wait until it is written
 
                     # Release the processor to other threads
-                    sched_yield()
+                    # sched_yield()
+                    time.sleep(0.0001)
 
                     # if we wait for a long time, log a message
                     if (time.monotonic() - start_time
