diff --git a/vllm/model_executor/models/eagle.py b/vllm/model_executor/models/eagle.py
index 3e4a5040b..8b73868f5 100644
--- a/vllm/model_executor/models/eagle.py
+++ b/vllm/model_executor/models/eagle.py
@@ -35,7 +35,7 @@ class DummyInputLayerNorm(nn.Module):
 
 class DummyOutputNorm(nn.Module):
 
-    def forward(self, x, residual):
+    def forward(self, x, residual=None):
         if residual is None:
             return x
         else:

