diff --git a/tests/unit_tests/data/__init__.py b/tests/unit_tests/data/__init__.py
deleted file mode 100644
index e69de29b..00000000
diff --git a/tests/unit_tests/data/__init__.py b/tests/unit_tests/data/__init__.py
new file mode 120000
index 00000000..fd5fa990
--- /dev/null
+++ b/tests/unit_tests/data/__init__.py
@@ -0,0 +1 @@
+/mnt/git_flagscale/patch_unpatch_pr/pr2_train_patch/FlagScale/flagscale/backends/Megatron-LM/tests/unit_tests/data/__init__.py
\ No newline at end of file
