From cd33c8caedf1a87a250ab5d03554397f8fc05293 Mon Sep 17 00:00:00 2001
From: brianlcy123 <brianlcy123@gmail.com>
Date: Tue, 24 Dec 2024 10:19:56 +0800
Subject: [PATCH] [kunlunxin] update mixtral patch for py39

---
 megatron/megatron/training/checkpointing.py | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/megatron/megatron/training/checkpointing.py b/megatron/megatron/training/checkpointing.py
index 6e58b317..7906ea88 100644
--- a/megatron/megatron/training/checkpointing.py
+++ b/megatron/megatron/training/checkpointing.py
@@ -1057,7 +1057,8 @@ def load_checkpoint(model, optimizer, opt_param_scheduler, load_arg='load', stri
             restore_modelopt_state(model, state_dict)
 
     # Model.
-    strict = False if args.retro_add_retriever else strict
+    #strict = False if args.retro_add_retriever else strict
+    strict = False
     if len(model) == 1:
         model[0].load_state_dict(state_dict['model'], strict=strict)
     else:
-- 
2.34.1
