diff --git a/vllm_mlu/vllm_mlu/distributed/device_communicators/mlu_communicator.py b/vllm_mlu/vllm_mlu/distributed/device_communicators/mlu_communicator.py
new file mode 100644
index 000000000..4e6d0dab2
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/distributed/device_communicators/mlu_communicator.py
@@ -0,0 +1,19 @@
+from typing import Optional
+
+import torch
+from torch.distributed import ProcessGroup
+
+from vllm.distributed.device_communicators.base_device_communicator import \
+    DeviceCommunicatorBase
+
+
+class MLUCommunicator(DeviceCommunicatorBase):
+
+    def __init__(self,
+                 cpu_group: ProcessGroup,
+                 device: Optional[torch.device] = None,
+                 device_group: Optional[ProcessGroup] = None,
+                 unique_name: str = ""):
+        super().__init__(cpu_group, device, device_group, unique_name)
+        # init device according to rank
+        self.device = torch.mlu.current_device()

