diff --git a/.gitmodules b/.gitmodules
index 0838ae01..a15fe18e 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -16,3 +16,9 @@
 [submodule "third_party/sglang"]
 	path = third_party/sglang
 	url = https://github.com/sgl-project/sglang.git
+[submodule "third_party/omniinfer"]
+	path = third_party/omniinfer
+	url = https://gitee.com/omniai/omniinfer.git
+[submodule "third_party/verl"]
+	path = third_party/verl
+	url = https://github.com/volcengine/verl.git

