diff --git a/vllm_mlu/vllm_mlu/engine/multiprocessing/__init__.py b/vllm_mlu/vllm_mlu/engine/multiprocessing/__init__.py
new file mode 100644
index 000000000..0271c1b84
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/engine/multiprocessing/__init__.py
@@ -0,0 +1,6 @@
+from enum import Enum
+
+
+class RPCSchedulerProfileRequest(Enum):
+    INIT_SCHEDULER_VIEW = 1
+    SAVE_SCHEDULER_VIEW = 2
\ No newline at end of file

