diff --git a/megatron/energon/metadataset/__init__.py b/megatron/energon/metadataset/__init__.py
new file mode 100644
index 000000000..a013183a1
--- /dev/null
+++ b/megatron/energon/metadataset/__init__.py
@@ -0,0 +1,17 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from megatron.energon.metadataset.dataset_loader import DatasetLoader
+from megatron.energon.metadataset.loader import load_dataset, prepare_metadataset
+from megatron.energon.metadataset.loader_interface import DatasetLoaderInterface
+from megatron.energon.metadataset.metadataset import Metadataset
+from megatron.energon.metadataset.metadataset_v2 import MetadatasetV2
+
+__all__ = [
+    "DatasetLoader",
+    "load_dataset",
+    "DatasetLoaderInterface",
+    "Metadataset",
+    "MetadatasetV2",
+    "prepare_metadataset",
+]

