diff --git a/tests/unit_tests/test_parallel_state.py b/tests/unit_tests/test_parallel_state.py
index ddee63877..de1abb95e 100644
--- a/tests/unit_tests/test_parallel_state.py
+++ b/tests/unit_tests/test_parallel_state.py
@@ -229,7 +229,13 @@ def test_different_initialize_order_consistency(src_tp_pp, ep_size):
 
     Utils.destroy_model_parallel()
 
-
+"""
+    Author: lizhiyu
+    Date: 2024-02-11
+    Action: 
+    Reason: This test always fails. The retated commit is
+            https://github.com/NVIDIA/Megatron-LM/commit/3e7ceda6b750a31fd3eb3a8bff3d811b4eabd289#diff-354da93a3dc4df1cd3362bba49b91f3145c6902f84839c4a8341079cbe7603a8
+"""
 @pytest.mark.parametrize(
     'src_tp_pp, ep_size',
     [((1, 2), 1), ((1, 4), 1), ((2, 2), 1), ((1, 2), 2), ((1, 4), 2), ((2, 2), 2)],
