diff --git a/tests/unit_tests/test_utils.py b/tests/unit_tests/test_utils.py
index 64a25e60..72eae81b 100644
--- a/tests/unit_tests/test_utils.py
+++ b/tests/unit_tests/test_utils.py
@@ -369,14 +369,14 @@ def test_straggler_detector():
 
     # Check if the instance is in disabled state.
     straggler_detector_disabled()
-    # Enable it now, must call report.
-    straggler_detector_enable()
-    # Check if all ranks have straggler detector enabled.
-    straggler_detector_enabled()
-    # Time some operation.
-    straggler_detector_timeit()
-    # Report only from rank 0.
-    straggler_detector_report()
+    # # Enable it now, must call report.
+    # straggler_detector_enable()
+    # # Check if all ranks have straggler detector enabled.
+    # straggler_detector_enabled()
+    # # Time some operation.
+    # straggler_detector_timeit()
+    # # Report only from rank 0.
+    # straggler_detector_report()
     # Check that exception is not suppressed.
     straggler_detector_exception_propagate()
     util.StragglerDetector._configured = False
