diff --git a/tests/scripts/functional_tests/test_all.sh b/tests/scripts/functional_tests/test_all.sh
index 1a204eac..f2606b15 100755
--- a/tests/scripts/functional_tests/test_all.sh
+++ b/tests/scripts/functional_tests/test_all.sh
@@ -2,20 +2,27 @@
 
 # Run each command and capture its return value
 commands=(
+    # For train
     "tests/scripts/functional_tests/test_task.sh --type train --task aquila"
-    # TODO: need fix
-    # "tests/scripts/functional_tests/test_task.sh --type train --task deepseek"
-    # "tests/scripts/functional_tests/test_task.sh --type train --task mixtral"
-    # "tests/scripts/functional_tests/test_task.sh --type train --task llava_onevision"
-    # for hetero-train
-    # "tests/scripts/functional_tests/test_task.sh --type hetero_train --task aquila"
-    # Add in the feature
-    "tests/scripts/functional_tests/test_task.sh --type inference --task deepseek"
+    "tests/scripts/functional_tests/test_task.sh --type train --task deepseek"
+    "tests/scripts/functional_tests/test_task.sh --type train --task mixtral"
+    "tests/scripts/functional_tests/test_task.sh --type train --task llava_onevision"
+    "tests/scripts/functional_tests/test_task.sh --type hetero_train --task aquila"
+    
+    # For inference
+    "tests/scripts/functional_tests/test_task.sh --type inference --task deepseek_r1_distill_qwen"
+    "tests/scripts/functional_tests/test_task.sh --type inference --task deepseek_r1_distill_qwen-flaggems"
+    "tests/scripts/functional_tests/test_task.sh --type inference --task deepseek_r1_distill_qwen-metax"
+    "tests/scripts/functional_tests/test_task.sh --type inference --task deepseek_r1_distill_qwen-flaggems-metax"
+    "tests/scripts/functional_tests/test_task.sh --type inference --task opi_llama3_1_instruct-metax"
+    "tests/scripts/functional_tests/test_task.sh --type inference --task opi_llama3_1_instruct-flaggems-metax"
     "tests/scripts/functional_tests/test_task.sh --type inference --task qwen3"
-    "tests/scripts/functional_tests/test_task.sh --type inference --task deepseek_flaggems"
-    "tests/scripts/functional_tests/test_task.sh --type inference --task qwen3_flaggems"
+    "tests/scripts/functional_tests/test_task.sh --type inference --task qwen3-flaggems"
+    "tests/scripts/functional_tests/test_task.sh --type inference --task qwen3-metax"
+    "tests/scripts/functional_tests/test_task.sh --type inference --task qwen3-flaggems-metax"
     # For serve
-    # "tests/scripts/functional_tests/test_task.sh --type serve --task base"
+    "tests/scripts/functional_tests/test_task.sh --type serve --task qwen2_5"
+    "tests/scripts/functional_tests/test_task.sh --type serve --task base"
 )
 
 for cmd in "${commands[@]}"; do

