diff --git a/vllm/v1/executor/multiproc_executor.py b/vllm/v1/executor/multiproc_executor.py
index 1d5175eb6..28a36acab 100644
--- a/vllm/v1/executor/multiproc_executor.py
+++ b/vllm/v1/executor/multiproc_executor.py
@@ -46,6 +46,9 @@ class MultiprocExecutor(Executor):
         # The child processes will send SIGUSR1 when unrecoverable
         # errors happen.
         def sigusr1_handler(signum, frame):
+            import traceback
+            logger.fatal("Singal received. Current stack:\n%s",
+                         "".join(traceback.format_stack(frame)))
             logger.fatal(
                 "MulitprocExecutor got fatal signal from worker processes, "
                 "shutting down. See stack trace above for root cause issue.")

