diff --git a/vllm_mlu/tools/build.property b/vllm_mlu/tools/build.property
new file mode 100644
index 000000000..347689733
--- /dev/null
+++ b/vllm_mlu/tools/build.property
@@ -0,0 +1,10 @@
+VLLM_VERSION=0.8.3+mlu0.8.0.pt26
+TORCH_MLU_OPS_VERSION=1.5.0+pt26
+CATCH_VERSION=1.26.1+torch2.6.0
+CNCL_VERSION=1.27.2-1
+CNNL_VERSION=2.1.0-1
+CNNLEXTRA_VERSION=2.0.3-1
+CNTOOLKIT_VERSION=4.1.2-1
+MLUOPS_VERSION=1.6.0-1
+TRITON_VERSION=3.2.0+mlu1.5.0
+XFORMERS_VERSION=0.0.24+mlu0.7.0.pt2.6

