diff --git a/tests/unit_tests/dist_checkpointing/test_optimizer.py b/tests/unit_tests/dist_checkpointing/test_optimizer.py
index df501366f..c64bb93bb 100644
--- a/tests/unit_tests/dist_checkpointing/test_optimizer.py
+++ b/tests/unit_tests/dist_checkpointing/test_optimizer.py
@@ -400,6 +400,7 @@ class TestDistributedOptimizer:
                     model, optimizer_A = setup_model_and_optimizer(
                         seed=2, tp=tp, pp=pp, initialize_fn=initialize_fn
                     )
+                
                 if test_step:
                     # Simulate "step" not set in some of the param groups on rank 0.
                     # TE FusedAdam may have "step" not set in some of the param groups on some ranks.
@@ -422,6 +423,7 @@ class TestDistributedOptimizer:
                     preprocess_common_state_dict_fn=preprocess_common_state_dict,
                 )
 
+                
                 # Get optimizer A param state
                 optim_param_state_A = optimizer_A.state_dict()
 
