diff --git a/tests/unit_tests/dist_checkpointing/utils.py b/tests/unit_tests/dist_checkpointing/utils.py
index 18d50e717..eced11189 100644
--- a/tests/unit_tests/dist_checkpointing/utils.py
+++ b/tests/unit_tests/dist_checkpointing/utils.py
@@ -138,6 +138,7 @@ def init_checkpointing_mock_args(args, ckpt_dir, fully_parallel=False):
     args.auto_detect_ckpt_format = False
     args.exit_on_missing_checkpoint = False
     args.finetune = False
+    args.finetune_with_optim = False
     args.consumed_train_samples = 0
     args.skipped_train_samples = 0
     args.consumed_valid_samples = 0

