diff --git a/tools/checkpoint/qwen2_5_vl/hf2mcore_qwen2.5_vl_convertor.sh b/tools/checkpoint/qwen2_5_vl/hf2mcore_qwen2.5_vl_convertor.sh
index cb0d0349..d76caa55 100644
--- a/tools/checkpoint/qwen2_5_vl/hf2mcore_qwen2.5_vl_convertor.sh
+++ b/tools/checkpoint/qwen2_5_vl/hf2mcore_qwen2.5_vl_convertor.sh
@@ -79,6 +79,28 @@ gqa_options=" \
 tie_option=" \
         --untie-embeddings-and-output-weights \
         "
+
+elif [ $MODEL_SIZE = 32B ]; then
+
+NUM_LAYERS=64
+HIDDEN_SIZE=5120
+NUM_ATTN_HEADS=40
+INTERMEDIATE_SIZE=27648
+NUM_KEY_VALUE_HEADS=8
+MAX_POSITION_EMBEDDINGS=128000
+EXTRA_VOCAB_SIZE=421
+RMS_NORM_EPS=1e-6
+
+gqa_options=" \
+		    --group-query-attention \
+		    --num-query-groups ${NUM_KEY_VALUE_HEADS}"
+
+
+tie_option=" \
+        --untie-embeddings-and-output-weights \
+        "
+
+
 elif [ $MODEL_SIZE = 72B ]; then
 
 NUM_LAYERS=80

