diff --git a/vllm_mlu/vllm_mlu/model_executor/models/mlu_models/__init__.py b/vllm_mlu/vllm_mlu/model_executor/models/mlu_models/__init__.py
new file mode 100644
index 000000000..e2723c1ac
--- /dev/null
+++ b/vllm_mlu/vllm_mlu/model_executor/models/mlu_models/__init__.py
@@ -0,0 +1,12 @@
+from vllm.model_executor.models.registry import ModelRegistry
+
+
+_MLU_VLLM_MODELS = {
+    "HunYuanForCausalLM": ("hunyuan", "HunYuanForCausalLM"),
+}
+
+for model_arch, (mod_relname, cls_name) in _MLU_VLLM_MODELS.items():
+    ModelRegistry.register_model(
+        model_arch=model_arch,
+        model_cls=f"vllm_mlu.model_executor.models.mlu_models.{mod_relname}:{cls_name}"
+    )
\ No newline at end of file

