diff --git a/megatron/energon/flavors/image_classification.py b/megatron/energon/flavors/image_classification.py
new file mode 100644
index 000000000..ac95269b6
--- /dev/null
+++ b/megatron/energon/flavors/image_classification.py
@@ -0,0 +1,40 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from typing import Optional
+
+import torch
+
+from megatron.energon.dataclass_slots import dataclass_slots
+from megatron.energon.epathlib.epath import EPath
+from megatron.energon.errors import warn_deprecated
+from megatron.energon.flavors.base_dataset import Sample
+from megatron.energon.flavors.webdataset import DefaultDecoderWebdatasetFactory
+
+
+@dataclass_slots
+class ImageClassificationSample(Sample):
+    """Sample type for classifying an image."""
+
+    #: The input image tensor in the shape (C, H, W)
+    image: torch.Tensor
+    #: The class label of the image
+    label: Optional[int] = None
+    #: The class label of the image
+    label_name: Optional[str] = None
+
+
+class ImageClassificationWebdataset(DefaultDecoderWebdatasetFactory[ImageClassificationSample]):
+    __sample_type__ = ImageClassificationSample
+
+    def __init__(self, path: EPath, **kwargs):
+        warn_deprecated(
+            f"{type(self)} is deprecated, use the default instead and set the sample_type:\n"
+            f"To convert, update your {path}/.nv-meta/dataset.yaml to:\n"
+            f"# remove top-level __module__ and __class__\n"
+            f"sample_type:\n"
+            f"  __module__: megatron.energon\n"
+            f"  __class__: {self.__sample_type__.__name__}\n"
+            f"# Keep the remaining content"
+        )
+        super().__init__(path, **kwargs)

