diff --git a/flagscale/train/mindspeed_patch/op_builder/__init__.py b/flagscale/train/mindspeed_patch/op_builder/__init__.py
new file mode 100644
index 00000000..e20f3ffe
--- /dev/null
+++ b/flagscale/train/mindspeed_patch/op_builder/__init__.py
@@ -0,0 +1,28 @@
+# from .fusion_attention_v2_builder import FusionAttentionV2OpBuilder
+# from .algorithm_builder import AlgorithmOpBuilder
+# from .adaptive_recomputing_builder import AdaptiveRecomputingPluggableAllocatorBuilder
+# from .npu_dropout_add_layer_norm_builder import NpuDropoutAddLayerNormOpBuilder
+from .atb_builder import AtbOpBuilder
+
+# from .swiglu_builder import SwigluOpBuilder
+# from .lcal_builder import LcalOpBuilder
+# from .rms_norm_builder import RmsNormOpBuilder
+# from .npu_grouped_mat_mul_all_reduce_builder import GroupedMatMulAllReduceOpBuilder
+# from .gmm_builder import GMMOpBuilder, GMMV2OpBuilder
+# from .quant_gmm_builder import QuantGMMOpBuilder
+# from .weight_quant_gmm_builder import WeightQuantGMMOpBuilder
+# from .ffn_builder import FFNOpBuilder
+# from .npu_mm_all_reduce_add_rms_norm_builder import MatmulAllReduceAddRmsNormOpBuilder
+# from .npu_inplace_mm_all_reduce_add_rms_norm_builder import InplaceMatmulAllReduceAddRmsNormOpBuilder
+# from .npu_rotary_position_embedding_builder import RotaryPositionEmbeddingOpBuilder
+# from .npu_moe_token_permute_builder import MoeTokenPermuteOpBuilder
+# from .npu_moe_token_unpermute_builder import MoeTokenUnpermuteOpBuilder
+# from .npu_ring_attention_update_builder import RingAttentionUpdateOpBuilder
+# from .npu_bmm_reduce_scatter_all_to_all_builder import BatchMatMulReduceScatterAlltoAllOpBuilder
+# from .npu_all_to_all_all_gather_bmm_builder import AllToAllAllGatherBatchMatMulOpBuilder
+# from .adaptive_cp_builder import AdaptiveCpOpBuilder
+from .matmul_add_builder import MatmulAddOpBuilder
+
+# from .groupmatmul_add_builder import GroupMatmulAddOpBuilder
+# from .fused_ema_adamw_builder import FusedEmaAdamWOpBuilder
+# from .smart_swap_builder import SmartSwapBuilder

