diff --git a/megatron/core/optimizer/optimizer_config.py b/megatron/core/optimizer/optimizer_config.py
index e9989cc14..f2e1ffa8b 100644
--- a/megatron/core/optimizer/optimizer_config.py
+++ b/megatron/core/optimizer/optimizer_config.py
@@ -180,6 +180,13 @@ class OptimizerConfig:
 
     config_logger_dir: str = ""
     """When non-empty, dumps entry-point configs to config_logger_dir"""
+    
+    ########## FlagScale Begin ##########
+    ################
+    # Grouped learning rate for multi-model training.
+    ################
+    vision_ration: float = 1.0
+    ########## FlagScale End ##########
 
     def __post_init__(self):
         """Check the validity of the config."""
