diff --git a/tests/functional_tests/test_cases/serve/utils/models/base_model.py b/tests/functional_tests/test_cases/serve/utils/models/base_model.py
index 672ee523..6ce021cc 100644
--- a/tests/functional_tests/test_cases/serve/utils/models/base_model.py
+++ b/tests/functional_tests/test_cases/serve/utils/models/base_model.py
@@ -1,16 +1,18 @@
 from util_models.util_model import fn
 
 
-def model_A(prompt):
-    result = prompt + "__add_model_A"
-    return fn(result)
+class ModelA:
+    def forward(self, prompt):
+        result = prompt + "__add_model_A"
+        return fn(result)
 
 
-def model_B(input_data):
-    res = input_data + "__add_model_B"
-    return res
+class ModelB:
+    def forward(self, input_data):
+        res = input_data + "__add_model_B"
+        return res
 
 
 if __name__ == "__main__":
     prompt = "introduce Bruce Lee"
-    print(model_A(prompt))
+    print(ModelA().forward(prompt))

