diff --git a/vllm/engine/arg_utils.py b/vllm/engine/arg_utils.py
index 6bb794177..9b17c4393 100644
--- a/vllm/engine/arg_utils.py
+++ b/vllm/engine/arg_utils.py
@@ -1176,6 +1176,9 @@ class EngineArgs:
             # global layers in interleaved sliding window models.
             sliding_window = model_config.get_sliding_window()
 
+        if getattr(model_config.hf_text_config, "model_type", None) == "Emu3":
+            sliding_window = None
+
         # Note(hc): In the current implementation of decode context
         # parallel(DCP), tp_size needs to be divisible by dcp_size,
         # because the world size does not change by dcp, it simply
