diff --git a/megatron/energon/flavors/webdataset/__init__.py b/megatron/energon/flavors/webdataset/__init__.py
new file mode 100644
index 000000000..2be29fef9
--- /dev/null
+++ b/megatron/energon/flavors/webdataset/__init__.py
@@ -0,0 +1,32 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from megatron.energon.av import AVData, AVDecoder
+from megatron.energon.flavors.webdataset.base_webdataset import BaseWebdatasetFactory
+from megatron.energon.flavors.webdataset.config import JOIN_INDEX_FILENAME, MAIN_FOLDER_NAME
+from megatron.energon.flavors.webdataset.default_decoder_webdataset import (
+    DefaultDecoderWebdatasetFactory,
+    ImageDecoder,
+)
+from megatron.energon.flavors.webdataset.default_generic_webdataset import (
+    DefaultGenericWebdatasetFactory,
+)
+from megatron.energon.flavors.webdataset.empty_dataset_error import EmptyDatasetError
+from megatron.energon.flavors.webdataset.joined_webdataset import JoinedWebdatasetFactory
+from megatron.energon.flavors.webdataset.metadata import WebdatasetMeta
+from megatron.energon.flavors.webdataset.standard_webdataset import StandardWebdatasetFactory
+
+__all__ = [
+    "BaseWebdatasetFactory",
+    "DefaultDecoderWebdatasetFactory",
+    "DefaultGenericWebdatasetFactory",
+    "EmptyDatasetError",
+    "ImageDecoder",
+    "JoinedWebdatasetFactory",
+    "StandardWebdatasetFactory",
+    "AVData",
+    "AVDecoder",
+    "WebdatasetMeta",
+    "MAIN_FOLDER_NAME",
+    "JOIN_INDEX_FILENAME",
+]

