diff --git a/megatron/energon/av/__init__.py b/megatron/energon/av/__init__.py
new file mode 100644
index 000000000..a81f23f32
--- /dev/null
+++ b/megatron/energon/av/__init__.py
@@ -0,0 +1,17 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+# SPDX-License-Identifier: BSD-3-Clause
+
+from megatron.energon.av.av_data import AVData
+from megatron.energon.av.av_decoder import (
+    AVDecoder,
+    AVWebdatasetDecoder,
+)
+from megatron.energon.av.utils import get_clips_uniform, get_single_frames_uniform
+
+__all__ = [
+    "AVDecoder",
+    "AVWebdatasetDecoder",
+    "AVData",
+    "get_clips_uniform",
+    "get_single_frames_uniform",
+]

