diff --git a/python/sglang/__init__.py b/python/sglang/__init__.py
index 509b145a..8fb6dd2b 100644
--- a/python/sglang/__init__.py
+++ b/python/sglang/__init__.py
@@ -1,8 +1,10 @@
+# Adapted from
+# https://github.com/sgl-project/sglang
+
 # SGLang public APIs
 
 # Frontend Language APIs
-from sglang.global_config import global_config
-from sglang.lang.api import (
+from sglang.api import (
     Engine,
     Runtime,
     assistant,
@@ -16,7 +18,6 @@ from sglang.lang.api import (
     get_server_info,
     image,
     select,
-    separate_reasoning,
     set_default_backend,
     system,
     system_begin,
@@ -26,26 +27,22 @@ from sglang.lang.api import (
     user_end,
     video,
 )
+from sglang.global_config import global_config
 from sglang.lang.backend.runtime_endpoint import RuntimeEndpoint
 from sglang.lang.choices import (
     greedy_token_selection,
     token_length_normalized,
     unconditional_likelihood_normalized,
 )
-
-# Lazy import some libraries
 from sglang.utils import LazyImport
 from sglang.version import __version__
 
+ServerArgs = LazyImport("sglang.srt.server_args", "ServerArgs")
 Anthropic = LazyImport("sglang.lang.backend.anthropic", "Anthropic")
 LiteLLM = LazyImport("sglang.lang.backend.litellm", "LiteLLM")
 OpenAI = LazyImport("sglang.lang.backend.openai", "OpenAI")
 VertexAI = LazyImport("sglang.lang.backend.vertexai", "VertexAI")
 
-# Runtime Engine APIs
-ServerArgs = LazyImport("sglang.srt.server_args", "ServerArgs")
-Engine = LazyImport("sglang.srt.entrypoints.engine", "Engine")
-
 __all__ = [
     "Engine",
     "Runtime",
@@ -60,7 +57,6 @@ __all__ = [
     "get_server_info",
     "image",
     "select",
-    "separate_reasoning",
     "set_default_backend",
     "system",
     "system_begin",
@@ -81,3 +77,8 @@ __all__ = [
     "global_config",
     "__version__",
 ]
+
+import torch._dynamo.config as pt_config
+
+# FX conflicts with Dynamo
+pt_config.error_on_nested_fx_trace = False
